cmd/gen-go-sample/enum.go

```diff
@@ -21,37 +21,55 @@ import (
 	"github.com/googleapis/gapic-generator-go/internal/pbinfo"
 )
 
+// enumFmt returns a function that transforms a protobuf enum value literal into it's corresponding Go const.
 func enumFmt(info pbinfo.Info, enum *descriptor.EnumDescriptorProto) func(*generator, string) (string, error) {
 	return func(g *generator, s string) (string, error) {
-		var element pbinfo.ProtoType = enum
-		var topElement pbinfo.ProtoType
-		var parts []string
-
-		for ; element != nil; element = info.ParentElement[element] {
-			parts = append(parts, element.GetName())
-			topElement = element
+		parts, impSpec, err := ancestors(info, enum, g)
+		if err != nil {
+			return "", err
 		}
-
 		if len(parts) > 1 {
 			// An enum is scoped using C++ rule. If MessageType contains EnumType,
 			// then elements are accessed as MessageType.Element, not MessageType.EnumType.Element.
 			//
 			// An exception to this rule is top-level enums, where len(parts)==1; they are qualified by
 			// EnumType.
-			parts = parts[1:]
+			parts = parts[:len(parts)-1]
 		}
 
-		// We made array in [child, parent, grandparent] order, we want grandparent fist.
-		for i := 0; i < len(parts)/2; i++ {
-			parts[i], parts[len(parts)-1-i] = parts[len(parts)-1-i], parts[i]
-		}
+		return impSpec.Name + "." + strings.Join(parts, "_") + "_" + s, nil
+	}
+}
 
-		impSpec, err := info.ImportSpec(topElement)
-		if err != nil {
-			return "", err
-		}
-		g.imports[impSpec] = true
+// enumType returns the Go Type name of a protobuf enum type.
+func enumType(info pbinfo.Info, enum *descriptor.EnumDescriptorProto, g *generator) (string, error) {
+	parts, impSpec, err := ancestors(info, enum, g)
+	if err != nil {
+		return "", err
+	}
+	return impSpec.Name + "." + strings.Join(parts, "_"), nil
+}
 
-		return impSpec.Name + "." + strings.Join(parts, "_") + "_" + s, nil
+// ancestors returns the names of all the types down to `enum`, included and the ImportSpec.
+// of the slice.
+func ancestors(info pbinfo.Info, enum *descriptor.EnumDescriptorProto, g *generator) ([]string, pbinfo.ImportSpec, error) {
+	var element pbinfo.ProtoType = enum
+	var topElement pbinfo.ProtoType
+	var parts []string
+
+	for ; element != nil; element = info.ParentElement[element] {
+		parts = append(parts, element.GetName())
+		topElement = element
+	}
+
+	// We made array in [child, parent, grandparent] order, we want grandparent fist.
+	for i := 0; i < len(parts)/2; i++ {
+		parts[i], parts[len(parts)-1-i] = parts[len(parts)-1-i], parts[i]
```

Comment:
can you put `len(parts)-1-i` in a variable since it is done twice (like in [SliceTricks](https://github.com/golang/go/wiki/SliceTricks#reversing) :wink:)

Reply:
Done